src/ethereum/utils.py

```diff
@@ -59,3 +76,354 @@ def ceil32(value: Uint) -> Uint:
         return value
     else:
         return value + ceiling - remainder
+
+
+def to_valid_address(data: bytes) -> Address:
+    """
+    Convert a byte stream to a valid address (20 bytes).
+
+    Parameters
+    ----------
+    data :
+        The string to be converted to bytes.
+
+    Returns
+    -------
+    byte_stream : `bytes`
+        The byte stream obtained from the passed string.
+    """
+    if len(data) == 20:
+        return data
+
+    if len(data) < 20:
+        return data.rjust(20, b"\x00")
+
+    return data[-20:]
```

Comment:
Any thoughts on making sure that we aren't truncating non-zero bytes?

Reply:
I would actually use an assert statement here to ensure that non-zero bytes are not getting truncated, but it seems that non-zero bytes can be truncated when forced to an address as per the yellow paper. To be more specific, I am referring to `BALANCE` EVM opcode from the yellow paper, as this opcode implementation also uses the same function. Let me know what you think, thanks.

Reply:
maybe this is two functions then?

- one for testing/utility that does `bytes -> bytes20`
- one for EVM operations that does `uint256 -> bytes20` with truncation/padding

The `BALANCE` opcode case should in theory be operating on a `uint256` popped off the stack and would need both padding and truncation.  The other use cases such as turning `b'\x01` into a full address for things like precompiles probably only needs to deal with padding and probably shouldn't truncate.